File: 141. Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == NULL) return false;
        ListNode *pslow = head;
        ListNode *pfast = head->next;
        while (pfast != NULL)
        {
            pslow = pslow->next;
            pfast = pfast->next;
            if (pfast != NULL)
                pfast = pfast->next;
            if (pslow == pfast)
                return true;
        }
        return false;
    }
};

File: 141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head:
            return False
        pslow = head
        pfast = head.next
        
        while pfast:
            pslow = pslow.next
            pfast = pfast.next
            if pfast:
                pfast = pfast.next
            if pslow == pfast:
                return True
        return False
